src/backend/commands/tablespace.c

```diff
@@ -1638,37 +1639,46 @@ assign_temp_tablespaces(const char *newval, void *extra)
 
 	/*
 	 * If check_temp_tablespaces was executed inside a transaction, then pass
-	 * the list it made to fd.c.  Otherwise, clear fd.c's list; we must be
-	 * still outside a transaction, or else restoring during transaction exit,
-	 * and in either case we can just let the next PrepareTempTablespaces call
-	 * make things sane.
+	 * the list it made to fd.c.  Otherwise, just let the next
+	 * PrepareTempTablespaces call make things sane.
 	 */
 	if (myextra)
 		SetTempTablespaces(myextra->tblSpcs, myextra->numSpcs);
 	else
-		SetTempTablespaces(NULL, 0);
+		SetTempTablespaces(NULL, -1);
```

Comment:
Why not
```suggestion
		SetTempTablespaces(NULL, 0);
```
?

Reply:
> A bug which prevented temp_tablespaces value from being updated for the very
first query after changing the GUC with gpconfig was fixed in
assign_temp_tablespaces(): in case of myextra being NULL, numSpcs variable
should be reset the to -1, and not 0, to allow PrepareTempTablespaces() to
re-parse the GUC value on the next transaction.

Reply:
I tried with proposal changes only of this one file and test passes!

Reply:
> A bug which prevented temp_tablespaces value from being updated for the very
> first query after changing the GUC with gpconfig was fixed in
> assign_temp_tablespaces()

Why do we need '\c' in tests?

Reply:
> '\c' in tests

Tests use a \c workaround to re-instantiate and re-send GUCs to backend sessions, because gpstop -u doesn't update GUCs on segments yet.